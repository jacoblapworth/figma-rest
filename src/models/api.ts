import {
  Branch,
  Component,
  ComponentSet,
  FrameOffset,
  FrameOffsetRegion,
  Region,
  Style,
  Vector,
} from './ast.js'

// 'files:read,file_variables:read,file_variables:write,file_comments:write,file_dev_resources:read,file_dev_resources:write,webhooks:write'
/** Scopes for personal access tokens and OAuth 2 tokens determine which endpoints can be accessed. */
export enum Scopes {
  /**
   * Read files, projects, users, versions, comments, components & styles, and webhooks.
   */
  'files:read',
  /**
   * Read variables in Figma file. Note: this is only available to members in Enterprise organizations.
   */
  'file_variables:read',
  /**
   * Read and write to variables in Figma file. Note: this is only available to members in Enterprise organizations.
   */
  'file_variables:write',
  /**
   * Post and delete comments and comment reactions in files.
   */
  'file_comments:write',
  /**
   * Read dev resources in files.
   */
  'file_dev_resources:read',
  /**
   * Read and write to dev resources in files.
   */
  'file_dev_resources:write',
  /** Create and manage webhooks. */
  'webhooks:write',
}

/** A comment or reply left by a user */
export interface Comment {
  /** Unique identifier for comment */
  id: string
  /** Positioning information of the comment. Includes information on the location of the comment pin, which is either the absolute coordinates on the canvas or a relative offset within a frame. If the comment is a region, it will also contain the region height, width, and position of the anchor in regards to the region. */
  client_meta: Vector | FrameOffset | Region | FrameOffsetRegion
  /** The file in which the comment lives */
  file_key: string
  /** If present, the id of the comment to which this is the reply */
  parent_id: string
  /** The user who left the comment */
  user: User
  /** The UTC ISO 8601 time at which the comment was left */
  created_at: string
  /** If set, the UTC ISO 8601 time the comment was resolved */
  resolved_at: string
  /** Only set for top level comments. The number displayed with the comment in the UI */
  order_id?: number
  /** Comment message */
  message: string
  /** An array of reactions to the comment */
  reactions: Reaction[]
}

/** A description of a user */
export interface User {
  /** Unique stable id of the user */
  id: string
  /** Name of the user */
  handle: string
  /** URL link to the user's profile image */
  img_url: string
  /** Email associated with the user's account. This will only be present on the /v1/me endpoint */
  email?: string
}

/** A version of a file */
export interface Version {
  /** Unique identifier for version */
  id: string
  /** The UTC ISO 8601 time at which the version was created */
  created_at: string
  /** The label given to the version in the editor */
  label: string
  /** The description of the version as entered in the editor */
  description: string
  /** The user that created the version */
  user: User
}

/** A Project can be identified by both the Project name, and the ProjectID. */
export interface Project {
  /** The ID of the project */
  id: number
  /** The name of the project */
  name: string
}

export type ReactionEmoji =
  | ':eyes:'
  | ':heart_eyes:'
  | ':heavy_plus_sign:'
  | ':+1:'
  | ':-1:'
  | ':joy:'
  | ':fire:'

export interface Reaction {
  /** The user who left the reaction */
  user: User
  /** The emoji type of reaction as a string */
  emoji: ReactionEmoji
  /** The UTC ISO 8601 time at which the reaction was left */
  created_at: string
}

export interface File {
  name: string
  role: string
  lastModified: string
  editorType: string
  thumbnailUrl: string
  version: String
  document: Node
  components: Record<string, Component>
  componentSets: Record<string, ComponentSet>
  schemaVersion: 0
  styles: Record<string, Style>
  mainFileKey: string
  branches: Branch[]
}
