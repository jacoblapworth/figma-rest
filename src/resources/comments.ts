import { Reaction, ReactionEmoji } from '../models/api.js'
import {
  FrameOffset,
  FrameOffsetRegion,
  Region,
  Vector,
} from '../models/ast.js'
import type { PaginatedResponse } from '../models/response.js'
import { BaseResource } from './base.js'

import { paginate } from './pagination.js'

type GetCommentsParams = {
  /**
   * If enabled, will return comments as their markdown equivalents when applicable.
   */
  as_md?: boolean
}

interface PostCommentsParams {
  /**
   * The text contents of the comment to post
   */
  message: string
  /**
   * The comment to reply to, if any. This must be a root comment, that is, you cannot reply to a comment that is a reply itself (a reply has a parent_id).
   */
  comment_id?: string
  /**
   * The position of where to place the comment.
   */
  client_meta?: Vector | FrameOffset | Region | FrameOffsetRegion
}

type CommentsResponse = {
  comments: Comment[]
}

type ReactionsParams = {
  cursor?: string
}

interface ReactionsResponse
  extends PaginatedResponse<{
    reactions: Reaction[]
  }> {}

type ReactParams = {
  emoji: ReactionEmoji
}

/**
 * Projects are collections of files that belong to a user, or more commonly a team.
 * They can be accessed by collaborators and can be made up of multiple files.
 *
 * @beta
 * @see https://www.figma.com/developers/api#comments
 */
export class CommentsResource extends BaseResource {
  /**
   * Gets a list of comments left on the file.
   */
  public async get(
    file: string,
    searchParams: GetCommentsParams
  ): Promise<Comment[]> {
    const response = await this._client.request
      .get(`v1/files/${file}/comments`, { searchParams })
      .json<CommentsResponse>()
    return response.comments
  }

  /**
   * Gets a list of comments left on the file. //TODO: Update docs
   */
  public async post(file: string, json: PostCommentsParams) {
    return this._client.request.post(`v1/files/${file}/comments`, { json })
  }

  /**
   * Deletes a specific comment. Only the person who made the comment is allowed to delete it.
   */
  public async delete(file_key: string, comment_id: string) {
    return this._client.request.delete(
      `v1/files/${file_key}/comments/${comment_id}`
    )
  }

  /**
   * Gets a paginated list of reactions left on the comment.
   */
  public reactions(
    file_key: string,
    comment_id: string,
    searchParams: ReactionsParams = {}
  ): AsyncIterableIterator<Reaction> {
    this._client.request.paginate.each
    return this._client.request.paginate<Reaction, ReactionsResponse>(
      `v1/files/${file_key}/comments/${comment_id}/reactions`,
      {
        resolveBodyOnly: true,
        responseType: 'json',
        searchParams,
        pagination: {
          transform: (response) => {
            if (response.request.options.responseType === 'json') {
              return response.body.reactions
            }

            return []
          },
          paginate,
        },
      }
    )
  }

  /**
   * Posts a new comment reaction on a file comment.
   */
  public async react(file_key: string, comment_id: string, json: ReactParams) {
    return this._client.request.post(
      `v1/files/${file_key}/comments/${comment_id}/reactions`,
      {
        json,
      }
    )
  }

  /**
   * Deletes a specific comment reaction. Only the person who made the comment reaction is allowed to delete it.
   */
  public async deleteReaction(
    file_key: string,
    comment_id: string,
    searchParams: ReactParams
  ) {
    return this._client.request.delete(
      `v1/files/${file_key}/comments/${comment_id}/reactions`,
      { searchParams }
    )
  }
}
