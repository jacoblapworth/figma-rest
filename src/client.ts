import got from 'got'
import type { Got } from 'got'

import { API_PROTOCOL, API_HOST, API_VER } from './config.js'
import { VariablesResource } from './resources/variables.js'
import { ProjectsResource } from './resources/projects.js'
import { CommentsResource } from './resources/comments.js'
import { FilesResource } from './resources/files.js'
import { UsersResource } from './resources/users.js'
import { VersionsResource } from './resources/versions.js'
import { ActivityLogsResource } from './resources/activityLogs.js'
import { PaymentsResource } from './resources/payments.js'
import { DevResourcesResource } from './resources/devResources.js'

type PersonalConfig = {
  personalAccessToken: string
}

type OAuthConfig = {
  oAuthToken: string
}

type SharedConfig = {
  apiVersion?: string
  protocol?: string
  host?: string
}

type ClientConfig = SharedConfig & (PersonalConfig | OAuthConfig)

export class FigmaClient {
  request: Got

  personalAccessToken?: string
  oAuthToken?: string

  activityLogs: ActivityLogsResource
  comments: CommentsResource
  devResources: DevResourcesResource
  files: FilesResource
  payments: PaymentsResource
  projects: ProjectsResource
  users: UsersResource
  variables: VariablesResource
  versions: VersionsResource

  constructor({
    apiVersion = API_VER,
    protocol = API_PROTOCOL,
    host = API_HOST,
    ...config
  }: ClientConfig) {
    this.request = got.extend({
      prefixUrl: new URL(`${protocol}://${host}`),
      responseType: 'json',
      resolveBodyOnly: true,
    })

    if ('personalAccessToken' in config) {
      this.personalAccessToken = config.personalAccessToken
      this.request = this.request.extend({
        headers: { 'X-Figma-Token': this.personalAccessToken },
      })
    }

    if ('oAuthToken' in config) {
      this.oAuthToken = config.oAuthToken
      this.request = this.request.extend({
        headers: { Authorization: `Bearer ${this.oAuthToken}` },
      })
    }

    this.activityLogs = new ActivityLogsResource(this)
    this.comments = new CommentsResource(this)
    this.devResources = new DevResourcesResource(this)
    this.files = new FilesResource(this)
    this.payments = new PaymentsResource(this)
    this.projects = new ProjectsResource(this)
    this.users = new UsersResource(this)
    this.variables = new VariablesResource(this)
    this.versions = new VersionsResource(this)
  }
}
