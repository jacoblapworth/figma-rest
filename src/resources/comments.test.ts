import { rest } from 'msw'
import { describe, expect, test } from 'vitest'
import { FigmaClient } from '../client.js'
import { Reaction } from '../models/api.js'
import { server } from '../__mocks__/server.js'
import { reactions } from '../__mocks__/fixtures/comments.js'

describe('comments', () => {
  const client = new FigmaClient({ personalAccessToken: 'abc' })

  test('reactions', async () => {
    server.use(
      rest.get(
        'https://api.figma.com/v1/files/:file/comments/:comment/reactions',
        (req, res, ctx) => {
          return res.once(
            ctx.json({ pagination: { next_page: '123' }, reactions })
          )
        }
      )
    )

    let items: Reaction[] = []

    for await (let reaction of client.comments.reactions('123', '456')) {
      items.push(reaction)
    }

    expect(items).toMatchSnapshot()
    expect(items.length).toBe(4)
  })
})
