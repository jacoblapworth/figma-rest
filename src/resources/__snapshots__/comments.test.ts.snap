// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comments > reactions 1`] = `
[
  {
    "created_at": "2023-02-16T22:43:52.679Z",
    "emoji": ":+1:",
    "user": {
      "handle": "Sue Huang",
      "id": "976296371487097187",
      "img_url": "https://s3-alpha.figma.com/profile/c56fc499-8e8a-4bba-bad1-fd7129e6032b",
    },
  },
  {
    "created_at": "2023-02-17T01:16:53.003Z",
    "emoji": ":+1:",
    "user": {
      "handle": "Finn Clark",
      "id": "606630225555448770",
      "img_url": "https://s3-alpha.figma.com/profile/edb3d868-a927-4838-8a55-57b45b51c4f9",
    },
  },
  {
    "created_at": "2023-02-16T22:43:52.679Z",
    "emoji": ":+1:",
    "user": {
      "handle": "Sue Huang",
      "id": "976296371487097187",
      "img_url": "https://s3-alpha.figma.com/profile/c56fc499-8e8a-4bba-bad1-fd7129e6032b",
    },
  },
  {
    "created_at": "2023-02-17T01:16:53.003Z",
    "emoji": ":+1:",
    "user": {
      "handle": "Finn Clark",
      "id": "606630225555448770",
      "img_url": "https://s3-alpha.figma.com/profile/edb3d868-a927-4838-8a55-57b45b51c4f9",
    },
  },
]
`;
